README-xena-trafficgen-integration.md

```diff
@@ -0,0 +1,138 @@
+
+* `--valkyrie2544-resolution_tput` : Specify resolution for throughput testing
+
+* `--src-macs`: MAC address that becomes source of first active entity and destination for second (if two exist).
+
+* `--dst-macs` : If specified, becomes destination of first active entity and source for second
+
+* `--src-ips` : IP address that becomes source of first active entity and destination for second (if two exist).
+
+* `--dst-ips` : If specified, becomes destination of first active entity (src-ips) and source for second
+
+* `--use-src-ip-flows` / `--use-dst-ip-flows` : Apply flows to both MAC and IP addresses (if enabled overrides MAC flows option). Invoking either src-ip or dst-ip, or both, yields the same result 
+        > Default : 1
+
+* `--use-src-mac-flows` / `--use-dst-mac-flows` : Apply flows to MAC addresses only (overridden if IP flows are enabled). Invoking either src-mac or dst-mac, or both, yields the same result
+        > Default: 1
+
+* `--xena_module` : Specify int corresponding to the Xena chassis module number.
+
+## Known Issues
+* SyntaxError from `trex_client.py`
+  * Open `/opt/trex/current/automation/trex_control_plane/interactive/trex/common/trex_client.py`
+  * Change line 1188 from: `self.conn.async.set_timeout_sec(timeout_sec)` to: `self.conn.async_.set_timeout_sec(timeout_sec)`
+
+* SyntaxError from `trex_conn.py`
+  * Open `/opt/trex/current/automation/trex_control_plane/interactive/trex/common/trex_conn.py`
+  * Change line 36 from: `self.async = TRexSubscriber(self.ctx, self.rpc)` to: `self.async_ = TRexSubscriber(self.ctx, self.rpc)`
+  * Change line 60 from: `self.async.disconnect()` to: `self.async_.disconnect()`
+  * Change line 89 from: `self.async.barrier()` to: `self.async_.barrier()`
+  * Change line 98 from: `return self.async.barrier(baseline = True)` to: `return self.async_.barrier(baseline = True)`
+  * Change line 110 from: `self.async.set_as_zombie()` to: `self.async_.set_as_zombie()`
+  * Change line 143 from: `return ( self.async.last_data_recv_ts is not None and ((time.time() - self.async.last_data_recv_ts) <= 3) )` to `return ( self.async_.last_data_recv_ts is not None and ((time.time() - self.async_.last_data_recv_ts) <= 3) )`
+  * Change line 189 from: `rc = self.async.connect()` to `rc = self.async_.connect()`
+
+* SyntaxError from `trex_global_stats.py`
+  * Open `/opt/trex/current/automation/trex_control_plane/interactive/trex/common/stats/trex_global_stats.py`
+  * Change line 81 from: `("async_util.", "{0}% / {1}".format( format_threshold(round_float(self.client.conn.async.monitor.get_cpu_util()), [85, 100], [0, 85]),` to: `("async_util.", "{0}% / {1}".format( format_threshold(round_float(self.client.conn.async_.monitor.get_cpu_util()), [85, 100], [0, 85]),`
+  * Change line 82 from: `format_num(self.client.conn.async.monitor.get_bps() / 8.0, suffix = "B/sec"))),` to: `format_num(self.client.conn.async_.monitor.get_bps() / 8.0, suffix = "B/sec"))),`
```

Comment:
Egads, I thought I had commented on this already - my apologies! Back at the end of August I wrote this README as a "setup from scratch" sort of guide, which presumed the user had nothing already installed or configured. I walked through the process using a fresh Fedora 32 install, and wrote down the bare minimum steps necessary just to get the Xena functionality working. 

(At the time of writing the patch) binary-search.py required that a user also install TRex in order to function (as mentioned in a previous comment). But even after installing TRex I still couldn't run binary-search.py due to several crash-inducing errors I'd never seen before. I still don't know why these problems suddenly popped up, but I eventually traced the errors back to the TRex library. Perhaps some update from Cisco unintentionally introduced a bug? I also can't rule out that the problem might have been self-inflicted from some flaw in my install/setup process.

Whatever the cause, I found that binary-search.py would stop crashing (at least enough so that the Xena bits worked) if I made a few tweaks to the TRex install. Hacky? Absolutely. But I lacked the necessary knowledge or skills to sort this problem out (or the time to develop such competencies). So I wrote those workarounds into the README in the hopes they'd be "good-enough" for folks needing to use Xena in the short term, at least while the TRex dependency issues  persisted.

TLDR; the TRex changes were just duct tape and will hopefully become unneeded/irrelevant/obsolete and removed as soon as possible.